src/json/stack.rs

```diff
@@ -0,0 +1,153 @@
+#[derive(PartialEq, Clone, Debug)]
+enum InternalStackElement {
+    InternalIndex(u32),
+    InternalKey(u16, u16), // start, size
+}
+
+impl Stack {
+    pub fn new() -> Stack {
+        Stack { stack: Vec::new(), str_buffer: Vec::new() }
+    }
+
+    /// Returns The number of elements in the Stack.
+    pub fn len(&self) -> usize { self.stack.len() }
+
+    /// Returns true if the stack is empty.
+    pub fn is_empty(&self) -> bool { self.stack.is_empty() }
+
+    /// Provides access to the StackElement at a given index.
+    /// lower indices are at the bottom of the stack while higher indices are
+    /// at the top.
+    pub fn get<'l>(&'l self, idx: usize) -> StackElement<'l> {
+        match self.stack[idx] {
+            InternalStackElement::InternalIndex(i) => StackElement::Index(i),
+            InternalStackElement::InternalKey(start, size) => {
+                StackElement::Key(str::from_utf8(
+                    &self.str_buffer[start as usize .. start as usize + size as usize]).unwrap())
+            }
+        }
+    }
+
+    /// Compares this stack with an array of StackElements.
+    pub fn is_equal_to(&self, rhs: &[StackElement]) -> bool {
+        if self.stack.len() != rhs.len() { return false; }
+        for i in 0..rhs.len() {
+            if self.get(i) != rhs[i] { return false; }
+        }
+        return true;
+    }
+
+    /// Returns true if the bottom-most elements of this stack are the same as
+    /// the ones passed as parameter.
+    pub fn starts_with(&self, rhs: &[StackElement]) -> bool {
+        if self.stack.len() < rhs.len() { return false; }
+        for i in 0..rhs.len() {
+            if self.get(i) != rhs[i] { return false; }
+        }
+        return true;
+    }
+
+    /// Returns true if the top-most elements of this stack are the same as
+    /// the ones passed as parameter.
+    pub fn ends_with(&self, rhs: &[StackElement]) -> bool {
+        if self.stack.len() < rhs.len() { return false; }
+        let offset = self.stack.len() - rhs.len();
+        for i in 0..rhs.len() {
+            if self.get(i + offset) != rhs[i] { return false; }
+        }
+        return true;
+    }
+
+    /// Returns the top-most element (if any).
+    pub fn top<'l>(&'l self) -> Option<StackElement<'l>> {
+        return match self.stack.last() {
+            None => None,
+            Some(&InternalStackElement::InternalIndex(i)) => Some(StackElement::Index(i)),
+            Some(&InternalStackElement::InternalKey(start, size)) => {
+                Some(StackElement::Key(str::from_utf8(
+                    &self.str_buffer[start as usize .. (start+size) as usize]
+                ).unwrap()))
+            }
+        }
+    }
+}
+
+// All the method of this trait are part of the Stack structure, 
+// but have to be exposed to other modules but not to the external 
+// users of the library. That'a why those methods are exposed via a trait.
+pub trait SemiPubMethods{
```

Comment:
Can you mark this trait as `#[doc(hidden)]`? Unfortunately rustdoc isn't so great about dealing with public-but-actually-private traits.